### PR TITLE
[WiP] Adds bulk actions for motions, users and so on

### PR DESCRIPTION
**Pull request by GabrielInTheWorld**
Currently implemented:

- Bulk actions for motions
- Bulk actions for users
- Bulk actions for agenda-items
- Bulk actions for assignments

TODO:
- Refactoring of `motion-import.service.ts`
- Maybe splitting user-repo into `user-repository` and `temporary-user-repository`?

### CODE REVIEW

File: client/src/app/core/core-services/action.service.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ export class ActionService {
     }
 
     private async _sendRequest<T>(request: RequestInfo): Promise<T> {
+        console.log('sendRequest', request);
```
Comment (GabrielInTheWorld) [inline client/src/app/core/core-services/action.service.ts:27]: Will be removed.

File: client/src/app/core/ui-services/base-filter-list.service.ts [hunk 2]
```diff
@@ -454,13 +454,14 @@ export abstract class BaseFilterListService<V extends BaseViewModel> {
...
     /**
-     * Checks if a given ViewBaseModel passes the filter.
+     * Checks if a given ViewBaseModel passes a filter.
      *
      * @param item Usually a view model
      * @param filter The filter to check
      * @returns true if the item is to be displayed according to the filter
      */
     private checkIncluded(item: V, filter: OsFilter): boolean {
+        console.log('checkIncluded', item, filter);
```
Comment (GabrielInTheWorld) [inline client/src/app/core/ui-services/base-filter-list.service.ts:464]: Will be removed, too.

**Comment by FinnStutzenstein** [comment]
> Maybe splitting user-repo into user-repository and temporary-user-repository?

No, since repos are bound to a collection and users and temporary users are the same collection. It might be a bit full in this repo due to the manifold actions, but the advantage is, that users in general are equal - there is no difference in normal handling between a temporary and non-temporary user.

File: client/src/app/core/repositories/agenda/agenda-item-repository.service.ts [hunk 3]
```diff
@@ -16,6 +16,8 @@ export interface AgendaListTitle {
     subtitle?: string;
 }
 
+export type AgendaItemType = 1 | 2 | 3;
```
Comment (FinnStutzenstein) [inline client/src/app/core/repositories/agenda/agenda-item-repository.service.ts:19]: Please move it to the `agenda-item.ts` and maybe use it in `ItemVisibilityChoices` as well as change the type of `number` in `AgendaItem`

File: client/src/app/core/repositories/agenda/agenda-item-repository.service.ts [hunk 4]
```diff
@@ -178,4 +180,24 @@ export class AgendaItemRepositoryService extends BaseRepositoryWithActiveMeeting
...
+        return this.sendBulkActionToBackend(AgendaItemAction.UPDATE, payload);
+    }
+
+    public bulkSetAgendaType(items: ViewAgendaItem[], agendaType: AgendaItemType): Promise<void> {
+        const payload: AgendaItemAction.UpdatePayload[] = items.map(item => ({ id: item.id, type: agendaType }));
+        return this.sendBulkActionToBackend(AgendaItemAction.UPDATE, payload);
+    }
+
+    public bulkRemoveItemsFromAgenda(items: ViewAgendaItem[]): Promise<void> {
+        const payload: AgendaItemAction.DeletePayload[] = items.map(item => ({ id: item.id }));
```
Comment (FinnStutzenstein) [inline client/src/app/core/repositories/agenda/agenda-item-repository.service.ts:200]: Just a notice. YOu often did not include the `DeletePayload` since it is mostly `Identifiable`. Maybe this is a good reason to include it. If you have time, please add the `DeleteAction` to each model.

File: client/src/app/core/repositories/base-repository.ts [hunk 5]
```diff
@@ -336,8 +336,9 @@ export abstract class BaseRepository<V extends BaseViewModel, M extends BaseMode
         });
     }
 
-    protected raiseError = (errorMessage: string | Error) => {
-        this.repositoryServiceCollector.errorService.showError(errorMessage);
+    protected raiseError = (error: Error) => {
+        this.repositoryServiceCollector.errorService.showError(error);
+        throw error;
```
Comment (FinnStutzenstein) [inline client/src/app/core/repositories/base-repository.ts:341]: Please do not rethrow here. If you rebase, the send*ActionToBackend Methods are adpoted to intercept, rather to catch, the error: https://github.com/OpenSlides/openslides-client/blob/master/client/src/app/core/repositories/base-repository.ts#L343

File: client/src/app/core/repositories/topics/topic-repository.service.ts [hunk 6]
```diff
@@ -91,14 +85,35 @@ export class TopicRepositoryService extends BaseIsAgendaItemAndListOfSpeakersCon
...
+    public duplicateMultipleTopics(agendaItems: ViewAgendaItem[]): Promise<void> {
+        const payload: TopicAction.CreatePayload[] = agendaItems.map(item =>
+            this.getCreatePayload(this.getDuplicatedTopic(item))
         );
+        return this.sendBulkActionToBackend(TopicAction.CREATE, payload);
+    }
+
+    private getDuplicatedTopic(topicAgendaItem: ViewAgendaItem): CreateTopic {
+        const viewTopic = topicAgendaItem.content_object as ViewTopic;
+        return new CreateTopic({
```
Comment (FinnStutzenstein) [inline client/src/app/core/repositories/topics/topic-repository.service.ts:101]: Still not a fan of create models. Please use the createAction interface for this. Maybe you are able to completely remove `CreateTopic` (it may be used during importing...).

File: client/src/app/site/agenda/components/agenda-item-list/agenda-item-list.component.ts [hunk 7]
```diff
@@ -42,6 +45,10 @@ import { hasListOfSpeakers } from '../../models/view-list-of-speakers';
     styleUrls: ['./agenda-item-list.component.scss']
 })
 export class AgendaItemListComponent extends BaseListViewComponent<ViewAgendaItem> implements OnInit {
+    public readonly AGENDA_TYPE_PUBLIC = 1;
```
Comment (FinnStutzenstein) [inline client/src/app/site/agenda/components/agenda-item-list/agenda-item-list.component.ts:48]: please also move to agenda_item.ts and unify the usage of these constants.

File: client/src/app/site/agenda/components/agenda-item-list/agenda-item-list.component.ts [hunk 8]
```diff
@@ -275,30 +282,32 @@ export class AgendaItemListComponent extends BaseListViewComponent<ViewAgendaIte
...
      */
-    public async setClosedSelected(closed: boolean): Promise<void> {
+    public async openSelectedItems(): Promise<void> {
         try {
-            for (const item of this.selectedRows) {
-                await this.repo.update({ closed: closed }, item);
-            }
+            await this.repo.bulkOpenItems(this.selectedRows);
+        } catch (e) {
+            this.raiseError(e);
```
Comment (FinnStutzenstein) [inline client/src/app/site/agenda/components/agenda-item-list/agenda-item-list.component.ts:300]: should not be necessary anymore since it is intercepted in the repos, right?

Comment (FinnStutzenstein) [reply client/src/app/site/agenda/components/agenda-item-list/agenda-item-list.component.ts:300]: Please also check other places where `raiseError` is used.

File: client/src/app/site/motions/modules/motion-workflow/components/workflow-list/workflow-list.component.ts [hunk 9]
```diff
@@ -80,8 +80,7 @@ export class WorkflowListComponent extends BaseListViewComponent<ViewMotionWorkf
                 {
                     idField: 'motion_workflow_ids'
                 }
-            ],
-            fieldset: []
+            ]
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/modules/motion-workflow/components/workflow-list/workflow-list.component.ts:83]: Please undo these changes - the `fieldset: []` clears required fields for the ViewMeeting. If it is not gives, the DEFAULT_FIELDSET will be used, which will currently fetch the `description` of the meeting, but it is not needed here.

File: client/src/app/site/motions/services/motion-multiselect.service.ts [hunk 10]
```diff
@@ -24,6 +31,14 @@ import { ViewMotion } from '../models/view-motion';
     providedIn: 'root'
 })
 export class MotionMultiselectService {
+    private get activeMeetingId(): ActiveMeetingIdService {
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/services/motion-multiselect.service.ts:34]: Make and resutn type does not fit. If the serviceCollector has the `activeMeetingIdService`, than there is a missing getter in one of the base classes (the base class, which required the service collector). Please add it there. Same for the action service.

File: client/src/app/site/motions/services/motion-multiselect.service.ts [hunk 11]
```diff
@@ -126,38 +130,27 @@ export class MotionMultiselectService {
...
-        } else {
-            throw new Error(
-                this.translate.instant('You cannot change the recommendation of motions in different workflows!')
+        // hacks custom Displayables from recommendations
+        // TODO: Recommendations should be an own class
+        const title = this.translate.instant('This will set the following recommendation for all selected motions:');
+        const choices: Displayable[] = this.workflowRepo
+            .getWorkflowStatesForMotions(motions)
+            .filter(workflowState => !!workflowState.recommendation_label)
+            .map(workflowState => ({
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/services/motion-multiselect.service.ts:139]: Why do you map this, it should not be needed? Recommendation is the name for a state with a recommendation_label; it is not a different model. (Same as the wording `amendment` for a motion with a lead motion assigned). If there are problems, please give me a call or so, so we can speak about it.

**Review by FinnStutzenstein — CHANGES_REQUESTED** [review]

File: client/src/app/core/repositories/topics/topic-repository.service.ts (hunk 6 above)
Comment (GabrielInTheWorld) [reply client/src/app/core/repositories/topics/topic-repository.service.ts:101]: Yes, I'm not, too. There are many changes, that are not pushed yet. Those "*-create-*"-files will be completely deleted. :)

Comment (FinnStutzenstein) [reply client/src/app/core/repositories/topics/topic-repository.service.ts:101]: Yeah!
